File: project_euler/p002_sum_fibonacci/p002_sum_fibonacci.ml
(* https://projecteuler.net/problem=2 *)

let fibonacci n =
  let rec aux fib last1 last2 =
    let current = last1 + last2 in
    if current > n then fib
    else aux (current::fib) current last1
  in aux [1; 1] 1 1

let sum_even list =
  let is_even n =
    n mod 2 = 0
  in List.fold_left (+) 0 (List.filter is_even list)

let () =
  sum_even (fibonacci 4000000) |> print_int;
  print_newline ();

File: project_euler/p002_sum_fibonacci/p002_sum_fibonacci.go
// https://projecteuler.net/problem=2

package main

func sum_even_fibonacci(n int) int {
	last1 := 1
	last2 := 1
	acc := 0
	current := last1 + last2
	for current < n {
		if current%2 == 0 {
			acc += current
		}
		last2 = last1
		last1 = current
		current = last1 + last2
	}
	return acc
}

func main() {
	print(sum_even_fibonacci(4000000))
}
